File: _ERP_/frontend/src/auth/permissions.js
/**
 * Role-based permission helpers
 * Roles: admin, hr_manager, project_manager, team_member
 * 
 * NOTE: These logic rules must exactly mirror app/auth/permissions.py
 */

export const SCRUM_ROLES = {
  PRODUCT_OWNER: 'product_owner',
  SCRUM_MASTER: 'scrum_master',
  TEAM_MEMBER: 'team_member',
};

export const ROLES = {
  ADMIN: 'admin',
  HR_MANAGER: 'hr_manager',
  PROJECT_MANAGER: 'project_manager',
  TEAM_MEMBER: 'team_member',
};

/**
 * Basic role check
 */
export function hasRole(user, allowedRoles) {
  if (!user?.role) return false;
  return allowedRoles.includes(user.role);
}

/**
 * Check if user is admin
 */
export function isAdmin(user) {
  return user?.role === ROLES.ADMIN;
}

/**
 * Admin or HR Manager can manage hiring
 */
export function canManageHiring(user) {
  return [ROLES.ADMIN, ROLES.HR_MANAGER].includes(user?.role);
}

/**
 * Admin or Project Manager can create/manage projects
 */
export function canManageProjects(user) {
  return [ROLES.ADMIN, ROLES.PROJECT_MANAGER].includes(user?.role);
}

/**
 * User can access project if: admin, manager, or member
 */
export function getUserScrumRole(user, project) {
  if (!user || !project || !project.members) return null;
  const member = project.members.find(m => m.user_id === user.id);
  return member ? member.scrum_role : null;
}

/**
 * User can access project if: admin, manager, or member
 */
export function canAccessProject(user, project) {
  if (!user) return false;
  if (isAdmin(user)) return true;
  
  return project?.members?.some(m => m.user_id === user.id);
}

/**
 * User can manage project if: admin, PM role, or PO/Scrum Master
 */
export function canManageProject(user, project) {
  if (!user) return false;
  if (isAdmin(user)) return true;
  if (user.role === ROLES.PROJECT_MANAGER) return true;
  
  const role = getUserScrumRole(user, project);
  return role === SCRUM_ROLES.PRODUCT_OWNER || role === SCRUM_ROLES.SCRUM_MASTER;
}

/**
 * Only managers can create tasks
 */
export function canCreateTask(user, project) {
  return canManageProject(user, project);
}

/**
 * Manager or assignee can change task status
 */
export function canEditTaskStatus(user, task, project) {
  if (!user) return false;
  const isManager = canManageProject(user, project);
  const isAssignee = task?.assigned_to?.some(a => (a.id === user.id || a === user.id));
  return isManager || isAssignee;
}

/**
 * Only project manager or admin can reassign tasks
 */
export function canReassignTask(user, project) {
  if (!user) return false;
  if (isAdmin(user)) return true;
  const role = getUserScrumRole(user, project);
  return role === SCRUM_ROLES.PRODUCT_OWNER || role === SCRUM_ROLES.SCRUM_MASTER;
}

/**
 * Only managers can delete tasks
 */
export function canDeleteTask(user, project) {
  return canManageProject(user, project);
}

// ── Legacy Aliases ───────────────────────────────────────────────────────────
// These maintain compatibility with existing components (Guard.jsx, etc.)
export const isProjectManager = canManageProject;
export const canEditTask = canEditTaskStatus;

/**
 * Role display labels
 */
export function formatRole(role) {
  if (!role) return '';
  return role.replace(/_/g, ' ');
}

File: _ERP_/backend/app/auth/permissions.py
"""
Centralized permission rules - single source of truth
All permission logic lives here. Import and use these helpers everywhere.
"""
from app.users.models import User, RoleEnum
from app.projects.models import Project, Task, ScrumRole


# ── Role Groups ───────────────────────────────────────────────────────────────

ADMIN_ROLES = (RoleEnum.admin,)
HR_ROLES = (RoleEnum.admin, RoleEnum.hr_manager)
PROJECT_MANAGER_ROLES = (RoleEnum.admin, RoleEnum.project_manager)
ALL_ROLES = (RoleEnum.admin, RoleEnum.hr_manager, RoleEnum.project_manager, RoleEnum.employee)


# ── Core Permission Checks ────────────────────────────────────────────────────

def is_admin(user: User) -> bool:
    """Check if user is admin"""
    return user.role == RoleEnum.admin


def can_manage_hiring(user: User) -> bool:
    """Admin or HR manager can manage hiring"""
    return user.role in HR_ROLES


def can_manage_projects(user: User) -> bool:
    """Admin or project manager can create/manage projects"""
    return user.role in PROJECT_MANAGER_ROLES


# ── Project Permissions ───────────────────────────────────────────────────────

def get_user_scrum_role(user: User, project: Project) -> ScrumRole | None:
    for m in project.members:
        if m.user_id == user.id:
            return m.scrum_role
    return None

def can_access_project(user: User, project: Project) -> bool:
    """User can access project if: admin or member"""
    return (
        is_admin(user)
        or any(m.user_id == user.id for m in project.members)
    )

def can_manage_project(user: User, project: Project) -> bool:
    """User can manage project if: admin, global project_manager, or PO/Scrum Master"""
    if is_admin(user):
        return True
    if user.role == RoleEnum.project_manager:
        return True
    role = get_user_scrum_role(user, project)
    return role in (ScrumRole.product_owner, ScrumRole.scrum_master)


# ── Task Permissions ──────────────────────────────────────────────────────────

def can_create_task(user: User, project: Project) -> bool:
    """Any member can create tasks."""
    return can_access_project(user, project)


def can_edit_task(user: User, task: Task, project: Project) -> bool:
    """Task details and status are editable by assignees or managers."""
    # Manager
    if can_manage_project(user, project):
        return True
    # Assignee
    return any(a.id == user.id for a in task.assigned_to)

def can_edit_task_status(user: User, task: Task, project: Project) -> bool:
    return can_edit_task(user, task, project)


def can_reassign_task(user: User, task: Task, project: Project) -> bool:
    """Only managers or the current assignee can reassign the task."""
    if can_manage_project(user, project):
        return True
    return any(a.id == user.id for a in task.assigned_to)


def can_delete_task(user: User, project: Project) -> bool:
    """Only managers can delete tasks."""
    return can_manage_project(user, project)
